src/vsr/replica.zig

```diff
@@ -1371,6 +1392,30 @@ pub fn ReplicaType(
             assert(!self.do_view_change_quorum);
             self.do_view_change_quorum = true;
 
+            const dvc_latest = DVCQuorum.dvc_latest(self.do_view_change_from_all_replicas);
+            if (dvc_latest.header.op > self.op_checkpoint_trigger()) {
```

Comment:
Hm, this doesn't feel right. `self.op_checkpoint_trigger()` can be very close to our op (eg, the next op might be checkpoint trigger), and it feels like this can reduce availability? 

Let's say we trigger a view change very close to checkpoint. Elected primary adds a couple of request to the pipeline just above the checkpoint, but those requests never commit. Still, if we do another view change on top of that, DVC from that primary would include those two uncommitted requests. 

Now that DVC would prevent any other primary from getting elected in this view, even if it has everything exctept those two uncommitted requests? I think even in the next view no different primary would be elected, because the poisoned primary with two outstanding uncommited requests would re-generate the same poisoned DVC for the new view?

Reply:
If the DVC quorum includes only one replica that is on the next checkpoint, then yes, that replica is the only one that can become the primary.

The view change will still eventually succeed (possibly after cycling a bit).

This is a deliberate tradeoff — the view change is (potentially) slower, but this is much simpler.
The alternative is that the lagging backup needs to catch-up to the latest checkpoint before it can even install the DVC headers to its log. Unless I'm missing something, that would be _really_ complicated.

Reply:
So, what worries me is that I think in my example there's no replicas on the next checkpoint? The two prepares which cause the primary to forfeit are uncommitted, and would actually be truncated by CTRL, should this primary decide not to forfeit.

Reply:
:thinking: yeah, I think this might actually formally break liveness? Just a single faulty node can prevent progress, if it faults in such a way that it is able to send DVC for every view, but isn't able to become primary itself (which, admitedly, not a too realistic failure mode). 

I think if we _actually_ checked for "is there any one node which checkpointed the next trigger", we would be fine --- if there's one node at a chcekpoint, then at least quorum of nodes would be at a checkpoint eventually. The problem here is that the `dvc_latest` node might not be at a checkpoint, but rather have some uncommitted ops.

Reply:
Ah I see, this is quite tricky.

If there are no ops from the next checkpoint in the DVC quorum, then we are fine. (This is true even if some replicas do have ops from the next checkpoint — if they aren't part of the DVC quorum, they will just be truncated per usual).

If there are ops from the next checkpoint in the DVC quorum, if the (now-retired) primary that prepared them keeps participating in view-changes, then it will eventually succeed.

The problem is that a lagging backup might receive a SV, then use those SV headers to participate in a DVC, even though it has not caught up. If the primary (that sent the SV) is now down, then those ops are part of the DVC quorum, but no replica can become the primary, they will just keep cycling. :thinking:

Reply:
Yeah, I think as-is this is both an availability and a liveness issue.